### PR TITLE
Implementing `repr` for JS

### PR DESCRIPTION
**Pull request by stisa**
~~**THIS IS NOT COMPLETE YET**~~

I have been working on implementing `repr` for the js backend, to close [#343](https://github.com/nim-lang/Nim/issues/343).  
I open this pr before finishing, to allow for feedback on the way I implemented this, as I am sure there are other ways to implement this that may be better/easier. I'm also not so sure about my approach here, as C made use of a lot of pointers arithmetic that I couldn't get to replicate in JS
I tried to match the output of `repr` in the C backend, but I took some liberties:
- strings aren't prefixed by a pointer
- seqs are prefixed by a `@` instead of a pointer
- `nil` strings are represented as `nil` instead of `""`. This is because strings in an empty array are represented as `nil` ( in C backend ), and also because I think an empty string is different than a `nil` string.

I also needed to use some `asm` statements, mostly for the iterators for array,enum,sets,objects , and to allow checking for undefined.

edit: the only iterator remaining is for sets, but I can remove it.

The code is not final, I intend to do a pass to format the code following the style guide, adjusting the comments I made, adding tests and generally cleaning it up.

Here's a test with output when run with c and when run with js: [gist](https://gist.github.com/stisa/207dbaa726a4dfb49007890b93016990)

State
------

- [x] tySet
- [x] tyArray
  - [~] tyArrayConstr should work, not sure how to test
- [x] tyTuple
- [x] tyObject
- [x] tyRef
- [x] tyPtr
- [x] tySequence
- [x] tyInt - only implemented the 64 one, but all other seem to work
  - tyInt8
  - tyInt16
  - tyInt32
  - tyInt64
- [x] tyUInt - uses the implementation for int, but all other seem to work
  - tyUInt8
  - tyUInt16
  - tyUInt32
  - ~~tyUInt64~~ 
- [x] tyFloat  - only implemented the 64 one, but all other seem to work
  - tyFloat32
  - tyFloat64
- [x] tyEnum
- [x] tyBool
- [x] tyChar
- [x] tyString
- [x] tyCString
- [x] tyRange
- [x] tyProc C seems to treat this a tuple? I treat it as just a pointer.
- [x] tyPointer ~~implemented, but I need to research why in the call to rawecho the element pointed to is passed as first argument~~

edit1: progress on tyPtr,tyRef,tyPointer. Moved to using pointers in reprAny. Will work on tyTuple next.
edit2: Well that was quick. All done, needs testing now.

### CODE REVIEW

**Comment by Araq** [comment]
Indeed. Such a large patch requires tests.

**Comment by stisa** [comment]
- Added a bunch of tests  in [tests/js/trepr](https://github.com/stisa/Nim/blob/51c4b35f25f7bed8b8adff8bd5774d23dc74e800/tests/js/trepr.nim)
- Use `JSON.stringify` to show something for invalid values and varargs/openarray ( not sure about this )

Also, is there a reason why ( with C backend ) a proc is a pointer when using repr on it, and a tuple (pointer,nil) if you assign the proc to a var?
```nim
proc test():int =
  echo "hello"
var 
  ptest = test
  nilproc : proc():int
  pp : proc():int = test

echo repr test # 000000000040165F
echo repr ptest # 000000000042F948

echo repr nilproc # [Field0 = nil,Field1 = nil]
echo repr pp # [Field0 = 000000000040165F,Field1 = nil]
```

**Comment by Araq** [comment]
> Also, is there a reason why ( with C backend ) a proc is a pointer when using repr on it, and a tuple (pointer,nil) if you assign the proc to a var while specifying the type?

Yes, .closure.

**Comment by stisa** [comment]
Oh didn't think of that. I don't think it's treated differently in js though, so I'm not sure if it has any value to replicate this in js.

**Comment by Araq** [comment]
>  I don't think it's treated differently in js though, so I'm not sure if it has any value to replicate this in js.

It's very different in JS since we just use JS's closure mechanism. 

> I'm not sure if it has any value to replicate this in js.

I agree, there is no need to replicate this in js.

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -1623,22 +1623,56 @@ proc genToArray(p: PProc; n: PNode; r: var TCompRes) =
     localError(x.info, "'toArray' needs an array literal")
   r.res.add(")")
 
+proc genReprAux(p: PProc, n: PNode, r: var TCompRes, magic:string, typ:Rope = nil) =
```
Comment (dom96) [inline compiler/jsgen.nim]: Please be consistent with your coding style. There is a lot of missing whitespace after the ``:`` in a lot of your code.

File: compiler/jsgen.nim [hunk 2]
```diff
@@ -1623,22 +1623,56 @@ proc genToArray(p: PProc; n: PNode; r: var TCompRes) =
...
+  useMagic(p,magic)
+  add(r.res, magic & "(")
+  var a: TCompRes
+
+  gen(p, n.sons[1], a)
+  if magic == "reprAny":
+    # the pointer argument in reprAny is expandend to 
+    # (pointedto, pointer), so we need to fill it
+    if a.address.isnil:
```
Comment (dom96) [inline compiler/jsgen.nim]: I know that Nim is style insensitive, but being consistent with the casing of identifiers is still important. This should be ``isNil``.

File: compiler/jsgen.nim [hunk 3]
```diff
@@ -1623,22 +1623,56 @@ proc genToArray(p: PProc; n: PNode; r: var TCompRes) =
...
+      add(r.res, a.res)
+      add(r.res, ", null") 
+    else:
+      add(r.res, "$1, $2" % [a.address, a.res])
+  else:
+    add(r.res,a.res)
```
Comment (dom96) [inline compiler/jsgen.nim]: Missing space here

File: compiler/jsgen.nim [hunk 4]
```diff
@@ -1623,22 +1623,56 @@ proc genToArray(p: PProc; n: PNode; r: var TCompRes) =
...
+
+  if not typ.isnil:
+    add(r.res,", ")
+    add(r.res,typ)
```
Comment (dom96) [inline compiler/jsgen.nim]: Bad casing and missing spaces around here.

File: compiler/jsgen.nim [hunk 5]
```diff
@@ -1623,22 +1623,56 @@ proc genToArray(p: PProc; n: PNode; r: var TCompRes) =
...
+
 proc genRepr(p: PProc, n: PNode, r: var TCompRes) =
   if p.target == targetPHP:
     localError(n.info, "'repr' not available for PHP backend")
     return
   let t = skipTypes(n.sons[1].typ, abstractVarRange)
   case t.kind
-  of tyInt..tyUInt64:
-    unaryExpr(p, n, r, "", "(\"\"+ ($1))")
+  of tyInt..tyInt64,tyUInt..tyUInt64:
```
Comment (dom96) [inline compiler/jsgen.nim]: Lots of missing spaces in this case statement. The ``..`` don't need spaces around it though.

File: lib/system/reprjs.nim [hunk 6]
```diff
@@ -6,18 +6,271 @@
...
 proc reprInt(x: int64): string {.compilerproc.} = return $x
+proc reprFloat(x: float): string {.compilerproc.} = 
+  # Js toString doesn't differentiate between 1.0 and 1,
+  # but we do.
+  if $x == $(x.int): $x & ".0"
+  else: $x
+
+proc reprPointer(p: pointer): string {.compilerproc.} = 
+  # Do we need to generate the full 8bytes ? In js a pointer is an int anyway
+  var tmp : int
```
Comment (dom96) [inline lib/system/reprjs.nim]: No space before the ``:``

File: lib/system/reprjs.nim [hunk 7]
```diff
@@ -6,18 +6,271 @@
...
+  if x: result = "true"
+  else: result = "false"
+
+proc isUndefined[T](x: T): bool {.inline.} = {.emit: "`result`= `x` === undefined;"}
 
 proc reprEnum(e: int, typ: PNimType): string {.compilerRtl.} =
-  if ntfEnumHole notin typ.flags:
-    if e <% typ.node.len:
-      return $typ.node.sons[e].name
+  if not typ.node.sons[e].isUndefined :
```
Comment (dom96) [inline lib/system/reprjs.nim]: no space before ``:``

File: lib/system/reprjs.nim [hunk 8]
```diff
@@ -6,18 +6,271 @@
...
+    result = $typ.node.sons[e].name
+  else:
+    result = $e & " (invalid data!)"
+  
+proc reprChar(x: char): string {.compilerRtl.} =
+  result = "\'"
+  case x
+  of '"': add result, "\\\""
```
Comment (dom96) [inline lib/system/reprjs.nim]: Put parenthesis around the ``add`` as well.

File: lib/system/reprjs.nim [hunk 9]
```diff
@@ -6,18 +6,271 @@
...
+    of '\10': add result, "\\10\"\n\""
+    of '\128' .. '\255', '\0'..'\9', '\11'..'\31':
+      add result, "\\" & reprInt(ord(c))
+    else:
+      add result, reprInt(ord(c)) # Not sure about this.
+  add result, "\""
+
+proc reprStr(s: string): string {.compilerRtl.} =
+  result = ""
+  if cast[pointer](s).isnil:
```
Comment (dom96) [inline lib/system/reprjs.nim]: Casing

File: lib/system/reprjs.nim [hunk 10]
```diff
@@ -6,18 +6,271 @@
...
+    # Also: c backend repr of a nil string is <pointer>"", but repr of an 
+    # array of string that is not initialized is [nil, nil, ...] ??
+    add result, "nil"
+    return
+  reprStrAux(result, s, s.len)
+
+proc addSetElem(result: var string, elem: int, typ: PNimType) =
+  # Dispatch each set element to the correct repr<Type> proc
+  case typ.kind
+  of tyEnum: add result, reprEnum(elem, typ)
```
Comment (dom96) [inline lib/system/reprjs.nim]: Parenthesis

File: lib/system/reprjs.nim [hunk 11]
```diff
@@ -6,18 +6,271 @@
...
+  of tyBool: add result, reprBool(bool(elem))
+  of tyChar: add result, reprChar(chr(elem))
+  of tyRange: addSetElem(result, elem, typ.base) # Note the base to advance towards the element type
+  of tyInt..tyInt64, tyUInt8, tyUInt16: add result, reprInt(elem)
+  else: # data corrupt --> inform the user
+    add result, " (invalid data!)"
+
+iterator SetKeys(s: int): int {.inline.} =
```
Comment (dom96) [inline lib/system/reprjs.nim]: Iterator/procedure names should begin with a lowercase letter.

File: lib/system/reprjs.nim [hunk 12]
```diff
@@ -6,18 +6,271 @@
...
+  # The type of s is a lie, but it's expected to be a set.
+  # Iterate over the JS object representing a set 
+  # and returns the keys as int.
+  var len: int
+  var yieldRes : int
+  var i : int = 0
```
Comment (dom96) [inline lib/system/reprjs.nim]: No space before ``:``

File: lib/system/reprjs.nim [hunk 13]
```diff
@@ -6,18 +6,271 @@
...
+  asm """
+  var setObjKeys = Object.getOwnPropertyNames(`s`);
+  `len` = setObjKeys.length
+  """
+  while i<len:
```
Comment (dom96) [inline lib/system/reprjs.nim]: Space before and after ``<``

File: lib/system/reprjs.nim [hunk 14]
```diff
@@ -6,18 +6,271 @@
...
+proc reprFloat(x: float): string {.compilerproc.} = 
+  # Js toString doesn't differentiate between 1.0 and 1,
+  # but we do.
+  if $x == $(x.int): $x & ".0"
+  else: $x
+
+proc reprPointer(p: pointer): string {.compilerproc.} = 
+  # Do we need to generate the full 8bytes ? In js a pointer is an int anyway
+  var tmp : int
+  asm """
```
Comment (dom96) [inline lib/system/reprjs.nim]: Is ``asm`` the correct thing to use here? Wouldn't ``emit`` make more sense?

File: lib/system/reprjs.nim [hunk 15]
```diff
@@ -6,18 +6,271 @@
...
+    asm "`yieldRes` = parseInt(setObjKeys[`i`],10);\n"
+    yield yieldRes
+    inc i
+
+proc reprSetAux(result: var string, s: int, typ: PNimType) =
+  add result, "{"
+  var first : bool = true
+  for el in SetKeys(s):
+    if first:
+      first  = false
```
Comment (dom96) [inline lib/system/reprjs.nim]: Too many spaces before ``=`` :)

File: lib/system/reprjs.nim [hunk 16]
```diff
@@ -6,18 +6,271 @@
...
+
+proc initReprClosure(cl: var ReprClosure) =
+  cl.recdepth = -1      # default is to display everything!
+  cl.indent = 0
+
+proc reprAux(result: var string, p: pointer, typ: PNimType, cl: var ReprClosure) 
+
+proc reprArray(a: pointer, typ: PNimType, 
+              cl: var ReprClosure): string {.compilerRtl.} =
+  var isnilarrorseq : bool
```
Comment (dom96) [inline lib/system/reprjs.nim]: Use camel case: ``isNilArrayOrSeq``

File: lib/system/reprjs.nim [hunk 17]
```diff
@@ -6,18 +6,271 @@
...
+    } else {`isnilarrorseq` = false
+    };
+    """
+  if typ.kind == tySequence and isnilarrorseq:
+    return "nil"
+
+  # We prepend @ to seq, the C backend prepends the pointer to the seq.
+  result = if typ.kind == tySequence: "@[" else: "["
+  var len: int = 0
+  var i : int = 0
```
Comment (dom96) [inline lib/system/reprjs.nim]: Spacing

File: lib/system/reprjs.nim [hunk 18]
```diff
@@ -6,18 +6,271 @@
...
+  for i in 0 .. < len:
+    if i>0 :
+      add result, ", "
+    asm "`dereffed`_Idx = `i`;\n" # advance the pointer
+    asm "`dereffed` = `a`[`dereffed`_Idx];\n" # point to element at index
+    reprAux(result, dereffed, typ.base, cl)  
+  
+  add result, "]"
+
+proc ispointedtonil(p: pointer): bool {.inline.}=
```
Comment (dom96) [inline lib/system/reprjs.nim]: camel case

File: lib/system/reprjs.nim [hunk 19]
```diff
@@ -6,18 +6,271 @@
...
+  """
+
+proc reprRef(result: var string, p: pointer, typ: PNimType,
+          cl: var ReprClosure) =
+  if p.ispointedtonil:
+    add result  , "nil"
+    return
+  add result, "ref " & reprPointer(p)
+  add result, " --> "
+  if not( typ.base.kind == tyArray):
```
Comment (dom96) [inline lib/system/reprjs.nim]: Use ``!=`` instead of ``not ... == ...``

File: lib/system/reprjs.nim [hunk 20]
```diff
@@ -6,18 +6,271 @@
...
+    asm """
+    if (`p` != null && `p`.length > 0) {
+      `p` = `p`[`p`_Idx];
+    }"""
+  reprAux(result, p, typ.base, cl)
+
+proc reprRecordAux(result: var string, o: pointer, typ: PNimType,cl: var ReprClosure) =
```
Comment (dom96) [inline lib/system/reprjs.nim]: spacing

File: lib/system/reprjs.nim [hunk 21]
```diff
@@ -6,18 +6,271 @@
...
+  var val: pointer = o
+  if typ.node.len == 0:
+    # if the object has only one field, len is 0  and sons is nil, the field is in node
+    let key: cstring =  typ.node.name
+    add result, $key & " = "
+    asm "`val` = `o`[`key`];\n"
+    reprAux(result, val, typ.node.typ, cl)
+  else:
+    # if the object has more than one field, sons is not nil and contains the fields.
+    for i in 0 .. < typ.node.len:
```
Comment (dom96) [inline lib/system/reprjs.nim]: No space after the ``<``

File: lib/system/reprjs.nim [hunk 22]
```diff
@@ -6,18 +6,271 @@
...
+      if first: first  = false
```
Comment (dom96) [inline lib/system/reprjs.nim]: Too many spaces before ``=``

File: lib/system/reprjs.nim [hunk 23]
```diff
@@ -6,18 +6,271 @@
...
+      else: add result, ",\n"
+
+      let key: cstring =  typ.node.sons[i].name
+      add result, $key & " = "
+      asm "`val` = `o`[`key`];\n" # access the field by name
+      reprAux(result, val, typ.node.sons[i].typ, cl)
+  add result, "]"
+
+proc reprRecord(o: pointer, typ: PNimType,cl: var ReprClosure): string {.compilerRtl.} =
```
Comment (dom96) [inline lib/system/reprjs.nim]: Spacing

File: lib/system/reprjs.nim [hunk 24]
```diff
@@ -6,18 +6,271 @@
...
+  result = ""
+  reprRecordAux(result, o, typ,cl)
```
Comment (dom96) [inline lib/system/reprjs.nim]: Spacing

File: lib/system/reprjs.nim [hunk 25]
```diff
@@ -6,18 +6,271 @@
...
+
+
+proc reprJSONStringify(p: int): string {.compilerRtl.}=
```
Comment (dom96) [inline lib/system/reprjs.nim]: Spacing

File: lib/system/reprjs.nim [hunk 26]
```diff
@@ -6,18 +6,271 @@
...
+  # As a last resort, use stringify
+  # We use this for tyOpenArray, tyVarargs while genTypeInfo is not implemented
+  var tmp: cstring
+  asm "`tmp` = JSON.stringify(`p`);"
+  result = $tmp
+
+proc reprAux(result: var string, p: pointer, typ: PNimType, 
+            cl: var ReprClosure) =
+  if cl.recdepth == 0:
```
Comment (dom96) [inline lib/system/reprjs.nim]: Camelcase?

File: lib/system/reprjs.nim [hunk 27]
```diff
@@ -6,18 +6,271 @@
...
+    add result, "..."
+    return 
+  dec(cl.recdepth)
+  case typ.kind
+  of tyInt..tyInt64,tyUInt..tyUInt64:
```
Comment (dom96) [inline lib/system/reprjs.nim]: Spacing

File: lib/system/reprjs.nim [hunk 28]
```diff
@@ -6,18 +6,271 @@
...
-    # ugh we need a slow linear search:
-    var n = typ.node
-    var s = n.sons
-    for i in 0 .. n.len-1:
-      if s[i].offset == e: return $s[i].name
-  result = $e & " (invalid data!)"
+    add result, "(invalid data!)" & reprJsonStringify(cast[int](p))
+  inc(cl.recdepth)
 
+proc reprAny(p: pointer, typ: PNimType): string {.compilerRtl.}=
```
Comment (dom96) [inline lib/system/reprjs.nim]: Spacing

File: tests/js/trepr.nim [hunk 29]
```diff
@@ -0,0 +1,340 @@
...
+    nc: int32 = -2147483000'i32
+    nd: int64 = -9223372036854775000'i64
+    ne: int = -1234567
+    pa: int8 = 120'i8
+    pb: int16 = 32700'i16
+    pc: int32 = 2147483000'i32
+    pd: int64 = 9223372036854775000'i64 
+    pe: int = 1234567
+  
+  doassert(repr(na) == "-120")
```
Comment (dom96) [inline tests/js/trepr.nim]: Camel case!

File: tests/js/trepr.nim [hunk 30]
```diff
@@ -0,0 +1,340 @@
...
+    nl = '\x0A'
+  
+  doassert(repr(a) == "'a'")
+  doassert(repr(b) == "'z'")
+  doassert(repr(one) == "'1'")
+  doassert(repr(nl) == "'\\10'")
+
+block strings:
+  let 
+    a:string = "12345"
```
Comment (dom96) [inline tests/js/trepr.nim]: Spacing

File: tests/js/trepr.nim [hunk 31]
```diff
@@ -0,0 +1,340 @@
...
+      a: int
+      b: float
+      c: string
+      d: OtherTuple
+    OtherTuple = tuple
+      a: bool
+      b: int8
+  
+  let
+    ot: OtherTuple = (a:true,b:120'i8)
```
Comment (dom96) [inline tests/js/trepr.nim]: spacing

File: tests/js/trepr.nim [hunk 32]
```diff
@@ -0,0 +1,340 @@
...
+    AnObj = object
+      a: int
+      b: float
+      c: OtherObj
+    OtherObj = object
+      a: bool
+      b: int8
+  let
+    oo: OtherObj = OtherObj(a:true,b:120'i8)
+    o: AnObj = AnObj(a:42, b:12.34,c:oo)
```
Comment (dom96) [inline tests/js/trepr.nim]: spacing

**Comment by stisa** [comment]
Thanks a lot for the review @dom96 ! I got a bit busy with other things, but I was planning on doing a pass for consistency, so your review helps a lot.
I'll have a look on asm vs emit.

Ps: I really need to look into reviving nimfix, I'm really messy at being consistent on my own.

**Comment by dom96** [comment]
Yes, nimfix would be great. In the long run we should have travis verify that the code style is consistent throughout.

**Pull request merged**